waffel/waffel.py

```diff
@@ -21,112 +22,165 @@
+                        objtype,
+                        ret['status']
+                    )
+                )
                 return None
             return ret['result']
 
         except requests.exceptions.RequestException as e:
-            print("fetching %s failed: %s" % (objtype, e))
+            logger.error("fetching %s failed: %s" % (objtype, e))
             return None
 
-    def get_events(self, year, track=None):
-        url = '%s?method=Event.detail&lang=en&year=%d' % (self.api_url, year)
+    def parse_date(self, dt):
+        return dateutil.parser.parse(dt)
+
+    def dt_to_epoch(self, dt):
+        # https://stackoverflow.com/a/11743262/1337474
+        return int((dt - datetime(1970, 1, 1)).total_seconds())
+
+    def dt_within(self, start, end):
+        now = datetime.now()
+        # now = datetime(2018, 2, 28, 18, 0)  # TODO: use/change this to test
+        if (
+            start > now + self.max_delta
+            or end < now - self.min_delta
+        ):
+            return False
+        return True
+
+    def make_event(self, start, end, title, subtitle, track_id):
+        # "start":     Startzeit in der gegebenen Zeitzone als HH:MM
+        # "startts":   Startzeit als unix epoch timestamp
+        # "end":       Endzeit in der gegebenen Zeitzone als HH:MM
+        # "endts":     Endzeit als unix epoch timestamp
+        # "title":     Titel des Event
+        # "subtitle":  Untertitel des Event (optional)
+        # "track":     discourse, music oder arts
+        #
+        # TODO: check, if timezone-naive is really OK
+        #       converting to timezone-aware datetimes shouldn't make a
+        #       difference.
+        return {
+            'start': start.strftime('%H:%M'),
+            'startts': self.dt_to_epoch(start),
+            'end': end.strftime('%H:%M'),
+            'endts': self.dt_to_epoch(end),
+            'title': title,
+            'subtitle': subtitle,
+            'track': self.track_map.get(track_id),
+        }
+
+    def get_events(self, track=None):
+        url = '%s?method=Event.detail&lang=en&year=%d' % (
+            self.api_url,
+            self.year
+        )
         if track:
             url = '%s&track=%s' % (url, track)
 
-        return self.__fetch_objects("events", url)
-
-    def get_locations(self, year, track=None):
-        url = '%s?method=Location.detail&lang=en&year=%d' % (self.api_url, year)
+        result = self.__fetch_objects("events", url)
+
+        ret = {}
+        for event in result:
+            start = self.parse_date(event['begin'])
+            end = self.parse_date(event['end'])
+            if not self.dt_within(start, end):
+                continue
+            location = self.location_map.get(event['location_id'], None)
+            if not location:
+                logger.warn(
+                    'location_id %s not found in location mapping' %
+                    event['location_id']
+                )
```

Comment:
Please only collect missing location id's here and print them later otherwise there will be potentially a lot of log lines like this.
Also please include the location names from EIS when you print the summary of missing locations (using an extra lookup on the api)

Reply:
done